Build and expose the OP's Entity Configuration

rmiccoli opened the pull request:
In the context of OpenID Federation 1.0 (following [draft 43](https://openid.net/specs/openid-federation-1_0.html)), the Entity Configuration (a signed JWT) of INDIGO IAM acting as OP has been defined and exposed at `/.well-known/openid-federation` endpoint.
The Entity Configuration is signed using an RSA key from the IAM keystore.

Example of Entity Configuration exposed by IAM:
_Decoded header_
```
{
  "kid": "rsa1",
  "typ": "entity-statement+jwt",
  "alg": "RS256"
}
```
_Decoded payload_
```
{
  "sub": "https://iam.example/",
  "metadata": {
    "openid_provider": {
      "response_types_supported": [
        "code",
        "token"
      ],
      "client_registration_types_supported": [
        "explicit"
      ],
      "federation_registration_endpoint": "https://iam.example/iam/api/oid-fed/client-registration",
      "jwks_uri": "https://iam.example/jwk",
      "subject_types_supported": [
        "public",
        "pairwise"
      ],
      "id_token_signing_alg_values_supported": [
        "HS256",
        "HS384",
        "HS512",
        "RS256",
        "RS384",
        "RS512",
        "ES256",
        "ES384",
        "ES512",
        "PS256",
        "PS384",
        "PS512",
        "none"
      ],
      "issuer": "https://iam.example/",
      "authorization_endpoint": "https://iam.example/authorize"
    },
     "federation_entity": {
      "logo_uri": "https://logo.example.com",
      "organization_name": "OID-FED",
      "contacts": [
        "iam-contact@example.com"
      ]
    }
  },
  "jwks": {
    "kty": "RSA",
    "e": "AQAB",
    "kid": "rsa1",
    "n": "4GRvJuFantVV3JdjwQOAkfREnwUFp2znRBTOIJhPamyH4gf4YlI5PQT79415NV4_HrWYzgooH5AK6-7WE-TLLGEAVK5vdk4vv79bG7ukvjvBPxAjEhQn6-Amln88iXtvicEGbh--3CKbQj1jryVU5aWM6jzweaabFSeCILVEd6ZT7ofXaAqan9eLzU5IEtTPy5MfrrOvWw5Q7D2yzMqc5LksmaQSw8XtmhA8gnENnIqjAMmPtRltf93wjtmiamgVENOVPdN-93Nd5w-pnMwEyoO6Q9JqXxV6lD6qBRxI7_5t4_vmVxcbbxcZbSAMoHqA2pbSMJ4Jcw-27Hct9jesLQ"
  },
  "iss": "https://iam.example/",
  "authority_hints": [
    "https://trust-anchor.sandbox.eosc.grnet.gr"
  ],
  "exp": 1758810387,
  "iat": 1758723987
}
```

rmiccoli committed:
Build and expose the OP's Entity Configuration

at /.well-known/openid-federation endpoint

rmiccoli committed:
Fix tests

rmiccoli committed:
Fix sonar issues and format

rmiccoli committed:
Add test

rmiccoli committed:
Fix code smell and format

rmiccoli committed:
Create an openid-federation spring profile

and make EC authority hints and expiration configurable

rmiccoli committed:
Add missing license

enricovianello commented:
> In the context of OpenID Federation 1.0 (following [draft 43](https://openid.net/specs/openid-federation-1_0.html)), the Entity Configuration (a signed JWT) of INDIGO IAM acting as OP has been defined and exposed at `/.well-known/openid-federation` endpoint. The Entity Configuration is signed using an RSA key from the IAM keystore.

Could you provide, in this description, an example of the metadata exposed with this PR?

rmiccoli commented:
> > In the context of OpenID Federation 1.0 (following [draft 43](https://openid.net/specs/openid-federation-1_0.html)), the Entity Configuration (a signed JWT) of INDIGO IAM acting as OP has been defined and exposed at `/.well-known/openid-federation` endpoint. The Entity Configuration is signed using an RSA key from the IAM keystore.
> 
> Could you provide, in this description, an example of the metadata exposed with this PR?

Done :)

rmiccoli committed:
Add federation_registration_endpoint

and federation_entity metadata

rmiccoli committed:
Change contacts type to List of strings

rmiccoli committed:
Add missing slash

enricovianello commented on iam-login-service/src/main/java/it/infn/mw/iam/core/oidc/EntityConfigurationBuilder.java:
```diff
@@ -0,0 +1,111 @@
+    metadata.put("openid_provider", opMetadata);
+    metadata.put("federation_entity", feMetadata);
+
+    JWTClaimsSet claims = new JWTClaimsSet.Builder().issuer(issuer)
```
if `issuer` has not a final slash it must be added to be coherent with current IAM logic

rmiccoli replied:
Actually, we should use `baseUrl`, which I believe is more appropriate in this context.

enricovianello commented on iam-login-service/src/main/java/it/infn/mw/iam/core/oidc/EntityConfigurationBuilder.java:
```diff
@@ -0,0 +1,111 @@
+    Map<String, Object> feMetadata = new HashMap<>();
+    feMetadata.put("contacts", List.of("iam-support@lists.infn.it"));
+    feMetadata.put("organization_name", iamProperties.getOrganisation().getName());
+    feMetadata.put("logo_uri", iamProperties.getLogo().getUrl());
```
The `logo_uri` should be an absolute URI. A relative path could be rejected by many clients.

rmiccoli replied:
I added a check so that if the `logo_uri` is not an absolute URI, it's not included in the federation entity metadata, as it's an optional field. Or would you prefer an error to be returned?

giacomini replied:
I usually prefer that if something is wrong in the configuration, the application doesn't start. Do we have other cases where we just ignore a wrong value?

rmiccoli replied:
I added an error in case the URI is defined and is not absolute.

enricovianello commented on iam-login-service/src/main/java/it/infn/mw/iam/core/oidc/EntityConfigurationBuilder.java:
```diff
@@ -0,0 +1,111 @@
+        .plusSeconds(openidFedProperties.getEntityConfiguration().getExpirationSeconds())))
+      .claim("jwks", jwks)
+      .claim("metadata", metadata)
+      .claim("authority_hints", openidFedProperties.getEntityConfiguration().getAuthorityHints())
```
we need to be 100% sure it's not empty

rmiccoli replied:
Added a check inside the constructor, so that, if the authority hints are empty or null, the application fails to start.

giacomini replied:
It's important that an empty authority_hints doesn't appear in the entity configuration. Does this check mean that you cannot even have a missing authority_hints in the entity configuration?

rmiccoli replied:
Yes, that's right. If we want IAM to also act as a Trust Anchor, then I must remove the check that authority hints is not null.

enricovianello commented on iam-login-service/src/main/java/it/infn/mw/iam/core/oidc/EntityConfigurationBuilder.java:
```diff
@@ -0,0 +1,111 @@
+  public EntityConfigurationBuilder(JWKSetKeyStore keyStore,
+      IamWellKnownInfoProvider wellKnownInfoProvider,
+      OpenidFederationProperties openidFedProperties, IamProperties iamProperties) {
+    this.wellKnownInfoProvider = wellKnownInfoProvider;
```
we need to correct the list of `id_token_signing_alg_values_supported` because we're saying (also in the OIDC well known) that `none` is supported. Check also that `none` is really not supported.

rmiccoli replied:
From https://openid.net/specs/openid-connect-discovery-1_0.html:
"_The value `none` MAY be supported but MUST NOT be used unless the Response Type used returns no ID Token from the Authorization Endpoint (such as when using the Authorization Code Flow)._ "

enricovianello commented on iam-login-service/src/main/java/it/infn/mw/iam/core/oidc/EntityConfigurationBuilder.java:
```diff
@@ -0,0 +1,111 @@
+
+    Map<String, Object> feMetadata = new HashMap<>();
+    feMetadata.put("contacts", List.of("iam-support@lists.infn.it"));
+    feMetadata.put("organization_name", iamProperties.getOrganisation().getName());
```
This feMetadata should be a yaml section so the values can be taken from a dedicated configuration section.

rmiccoli replied:
Done

enricovianello commented on iam-login-service/src/main/java/it/infn/mw/iam/core/web/federation/EntityConfigurationEndpoint.java:
```diff
@@ -0,0 +1,54 @@
+  }
+
+  @GetMapping(value = "/.well-known/openid-federation",
+      produces = "application/entity-statement+jwt")
```
we should manage also the headers to manage client caching of this response as we have done for the well known OIDC endpoint

rmiccoli replied:
Done

enricovianello commented on iam-login-service/src/main/java/it/infn/mw/iam/core/web/federation/EntityConfigurationEndpoint.java:
```diff
@@ -0,0 +1,54 @@
+  public ResponseEntity<byte[]> getEntityConfiguration() throws ParseException, JOSEException {
+    String jsonKeys = jwkController.getJwk().getBody();
+    Map<String, Object> jwks = JSONObjectUtils.parse(jsonKeys);
+    String ecJwt = builder.buildEntityConfiguration(jwks);
```
build is better as name

giacomini replied:
maybe change the name of the object, e.g. to entityConfigurationBuilder

rmiccoli replied:
Done

enricovianello commented on iam-login-service/src/main/java/it/infn/mw/iam/core/web/federation/EntityConfigurationEndpoint.java:
```diff
@@ -0,0 +1,54 @@
+      produces = "application/entity-statement+jwt")
+  public ResponseEntity<byte[]> getEntityConfiguration() throws ParseException, JOSEException {
+    String jsonKeys = jwkController.getJwk().getBody();
+    Map<String, Object> jwks = JSONObjectUtils.parse(jsonKeys);
```
this part could be moved inside the constructor

rmiccoli replied:
Done

enricovianello commented on iam-login-service/src/main/java/it/infn/mw/iam/core/web/federation/EntityConfigurationEndpoint.java:
```diff
@@ -0,0 +1,54 @@
+    String jsonKeys = jwkController.getJwk().getBody();
+    Map<String, Object> jwks = JSONObjectUtils.parse(jsonKeys);
+    String ecJwt = builder.buildEntityConfiguration(jwks);
+    return ResponseEntity.ok().body(ecJwt.getBytes(StandardCharsets.US_ASCII));
```
We can return directly byte[]

rmiccoli replied:
I think we need `ResponseEntity` to manage the headers for client caching of the `.well-known/openid-federation` response (as done for `/jwk` endpoint [here](https://github.com/indigo-iam/iam/blob/develop/iam-login-service/src/main/java/it/infn/mw/iam/core/web/jwk/IamJWKSetPublishingEndpoint.java#L55)).

enricovianello reviewed (changes requested):
Changes requested

rmiccoli committed:
Apply requested changes

- make federation_entity claims configurable in yml file
- manage header for client caching
- check authority_hints are not empty at startup
- check logo uri is an absolute uri
- retrieve public key in the constructor as for private key
- other cosmetic fixes

rmiccoli committed:
Change federation registration endpoint name

and fix code smells

giacomini commented on iam-login-service/src/main/java/it/infn/mw/iam/core/oidc/EntityConfigurationBuilder.java:
```diff
@@ -0,0 +1,135 @@
+    if (contacts != null && !contacts.isEmpty()) {
+      feMetadata.put("contacts", contacts);
+    }
+    if (logoUri != null && !logoUri.isBlank() && URI.create(logoUri).isAbsolute()) {
```
For how you manage the `logoUri` (see other discussion), `isAbsolute()` is always true, right? and can it be blank?

rmiccoli replied:
The validation is performed within the `build()` method (not within the constructor). Since `logoUri` is optional, it can be empty, but if defined, we check that it is absolute. If not, now we generate an error.
Would you prefer this check to be performed within the constructor so that if `logoUri` is not "well-defined", the application doesn't start?

rmiccoli replied:
It has been moved within the constructor.

rmiccoli committed:
Refactor builder
